### ISSUE TITLE
C++ STL class not found when compiling  4.1.0 "std::numeric_limits does not exist"

### ISSUE THREAD

**Issue by makingglitches** [bazelbuild/bazel#13613]
Hello.
Building this on Fedora 34.
OpenJDK version is 11.

Build of dist package kicks off erroring out at the following.

INFO: From Compiling src/google/protobuf/message_lite.cc [for host]:
In file included from /usr/include/string.h:519,
                 from external/com_google_protobuf/src/google/protobuf/stubs/port.h:39,
                 from external/com_google_protobuf/src/google/protobuf/stubs/macros.h:34,
                 from external/com_google_protobuf/src/google/protobuf/stubs/common.h:46,
                 from external/com_google_protobuf/src/google/protobuf/message_lite.h:45,
                 from external/com_google_protobuf/src/google/protobuf/message_lite.cc:36:
In function 'void* memcpy(void*, const void*, size_t)',
    inlined from 'google::protobuf::uint8* google::protobuf::io::EpsCopyOutputStream::WriteRaw(const void*, int, google::protobuf::uint8*)' at external/com_google_protobuf/src/google/protobuf/io/coded_stream.h:699:16,
    inlined from 'virtual google::protobuf::uint8* google::protobuf::internal::ImplicitWeakMessage::_InternalSerialize(google::protobuf::uint8*, google::protobuf::io::EpsCopyOutputStream*) const' at external/com_google_protobuf/src/google/protobuf/implicit_weak_message.h:85:28,
    inlined from 'bool google::protobuf::MessageLite::SerializePartialToZeroCopyStream(google::protobuf::io::ZeroCopyOutputStream*) const' at external/com_google_protobuf/src/google/protobuf/message_lite.cc:419:30:
/usr/include/bits/string_fortified.h:29:33: warning: 'void* __builtin___memcpy_chk(void*, const void*, long unsigned int, long unsigned int)' specified size between 18446744071562067968 and 18446744073709551615 exceeds maximum object size 9223372036854775807 [-Wstringop-overflow=]
   29 |   return __builtin___memcpy_chk (__dest, __src, __len,
      |          ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
   30 |                                  __glibc_objsize0 (__dest));
      |                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~
ERROR: /Data/Documents/CensusProject/gdallib/src/bazel-dist/third_party/ijar/BUILD:47:11: Compiling third_party/ijar/zlib_client.cc failed: (Exit 1): gcc failed: error executing command 
  (cd /tmp/bazel_iTrBEl3K/out/execroot/io_bazel && \
  exec env - \
    PATH=/home/john/perl5/bin:/usr/bin/pycharm-community-2021.1.1/bin:/home/john/perl5/bin:/home/john/.local/bin:/home/john/bin:/usr/bin/pycharm-community-2021.1.1/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/home/john/.dotnet/tools:/usr/local/lib/nodejs/node-v16.0.0-linux-x64/bin:/usr/local/lib/nodejs/node-v16.0.0-linux-x64/bin \
    PWD=/proc/self/cwd \
  /usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections -fdata-sections '-std=c++0x' -MD -MF bazel-out/k8-opt/bin/third_party/ijar/_objs/zlib_client/zlib_client.d '-frandom-seed=bazel-out/k8-opt/bin/third_party/ijar/_objs/zlib_client/zlib_client.o' -iquote . -iquote bazel-out/k8-opt/bin -isystem third_party/zlib -isystem bazel-out/k8-opt/bin/third_party/zlib -fno-canonical-system-headers -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -c third_party/ijar/zlib_client.cc -o bazel-out/k8-opt/bin/third_party/ijar/_objs/zlib_client/zlib_client.o)
Execution platform: //:default_host_platform
In file included from third_party/ijar/zlib_client.cc:20:
./third_party/ijar/zlib_client.h:64:46: error: 'numeric_limits' is not a member of 'std'
   64 |   static const size_t MAX_BUFFER_SIZE = std::numeric_limits<int32_t>::max();
      |                                              ^~~~~~~~~~~~~~
./third_party/ijar/zlib_client.h:64:68: error: expected primary-expression before '>' token
   64 |   static const size_t MAX_BUFFER_SIZE = std::numeric_limits<int32_t>::max();
      |                                                                    ^
./third_party/ijar/zlib_client.h:64:71: error: '::max' has not been declared; did you mean 'std::max'?
   64 |   static const size_t MAX_BUFFER_SIZE = std::numeric_limits<int32_t>::max();
      |                                                                       ^~~
      |                                                                       std::max
In file included from /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/algorithm:62,
                 from third_party/ijar/zlib_client.cc:16:
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/stl_algo.h:3467:5: note: 'std::max' declared here
 3467 |     max(initializer_list<_Tp> __l, _Compare __comp)
      |     ^~~
Target //src:bazel_nojdk failed to build
INFO: Elapsed time: 445.173s, Critical Path: 39.41s
INFO: 401 processes: 17 internal, 384 local.
FAILED: Build did NOT complete successfully

ERROR: Could not build Bazel

**Comment by makingglitches** [comment]
This is literally happening because this is not in any file named limits.h on my entire system !
Which is fedora 34 !

**Comment by makingglitches** [comment]
https://stackoverflow.com/questions/19165934/where-is-the-limit-h-file-that-defines-the-template-numeric-limits

That is ridiculous.

The error is fixed when changing the line 

#include <limits.h>

in the the file  third_party/ijar/zlib_client.h

to

#include <limits>

this however is not the end of the errors.

**Comment by makingglitches** [comment]
Gets better. recompilation results in:

> 
> ./third_party/ijar/zlib_client.h:52:19: error: 'PATH_MAX' was not declared in this scope
>    52 |   char errmsg[4 * PATH_MAX];
>       |                   ^~~~~~~~
> ./third_party/ijar/zlib_client.h:52:19: note: the macro 'PATH_MAX' had not yet been defined
> In file included from /usr/include/bits/local_lim.h:38,
>                  from /usr/include/bits/posix1_lim.h:161,
>                  from /usr/include/limits.h:195,
>                  from /usr/lib/gcc/x86_64-redhat-linux/11/include/limits.h:203,
>                  from /usr/lib/gcc/x86_64-redhat-linux/11/include/syslimits.h:7,
>                  from /usr/lib/gcc/x86_64-redhat-linux/11/include/limits.h:34,
>                  from /Data/Documents/CensusProject/gdallib/src/compiled/include/zconf.h:418,
>                  from /Data/Documents/CensusProject/gdallib/src/compiled/include/zlib.h:34,
>                  from third_party/ijar/zlib_client.cc:21:
> **/usr/include/linux/limits.h:13: note: it was later defined here**
>    13 | #define PATH_MAX        4096    /* # chars in a path name including nul */
> 

The source needs both.   Waiting on the end results of recompilation.

**Comment by makingglitches** [comment]
The two don't play nice.
when I include limits and limits.h in the same file the original error returns.

**Comment by makingglitches** [comment]
Not a c++ guy btw so. just saying.

**Comment by makingglitches** [comment]
moving some things around from zlib_client.cc to zlib_client.h  solved the compilation issue.  However then another error was encountered seemingly related to the same thing.

If I had to clasify this issue I'd say that it relates to differing version of the STL lib. God knows what version its compiling against since literally this is the first time I've compiled a package on Linux and had these specific kind of issues.  

Kind of annoying since i'm only trying to build this to get TensorFlow built for fedora.

>       |                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~
> ERROR: /tmp/bazel_8MmdvFOC/out/external/com_google_absl/absl/synchronization/BUILD.bazel:30:11: Compiling absl/synchronization/internal/graphcycles.cc failed: (Exit 1): gcc failed: error executing command 
>   (cd /tmp/bazel_8MmdvFOC/out/execroot/io_bazel && \
>   exec env - \
>     LD_LIBRARY_PATH=/Data/Documents/CensusProject/gdallib/src/compiled/lib/ \
>     PATH=/home/john/perl5/bin:/usr/bin/pycharm-community-2021.1.1/bin:/home/john/perl5/bin:/home/john/.local/bin:/home/john/bin:/usr/bin/pycharm-community-2021.1.1/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/home/john/.dotnet/tools:/usr/local/lib/nodejs/node-v16.0.0-linux-x64/bin:/usr/local/lib/nodejs/node-v16.0.0-linux-x64/bin:/lib/jvm/java-11-openjdk-11.0.11.0.9-4.fc34.x86_64/bin \
>     PWD=/proc/self/cwd \
>   /usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections -fdata-sections '-std=c++0x' -I/Data/Documents/CensusProject/gdallib/src/compiled/include -MD -MF bazel-out/k8-opt/bin/external/com_google_absl/absl/synchronization/_objs/graphcycles_internal/graphcycles.d '-frandom-seed=bazel-out/k8-opt/bin/external/com_google_absl/absl/synchronization/_objs/graphcycles_internal/graphcycles.o' -D__CLANG_SUPPORT_DYN_ANNOTATION__ -iquote external/com_google_absl -iquote bazel-out/k8-opt/bin/external/com_google_absl -Wall -Wextra -Wcast-qual -Wconversion-null -Wmissing-declarations -Woverlength-strings -Wpointer-arith -Wunused-local-typedefs -Wunused-result -Wvarargs -Wvla -Wwrite-strings -Wno-missing-field-initializers -Wno-sign-compare -fno-canonical-system-headers -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -c external/com_google_absl/absl/synchronization/internal/graphcycles.cc -o bazel-out/k8-opt/bin/external/com_google_absl/absl/synchronization/_objs/graphcycles_internal/graphcycles.o)
> Execution platform: //:default_host_platform
> external/com_google_absl/absl/synchronization/internal/graphcycles.cc: In member function 'void absl::lts_2020_02_25::synchronization_internal::GraphCycles::RemoveNode(void*)':
> external/com_google_absl/absl/synchronization/internal/graphcycles.cc:451:26: error: 'numeric_limits' is not a member of 'std'
>   451 |   if (x->version == std::numeric_limits<uint32_t>::max()) {
>       |                          ^~~~~~~~~~~~~~
> external/com_google_absl/absl/synchronization/internal/graphcycles.cc:451:49: error: expected primary-expression before '>' token
>   451 |   if (x->version == std::numeric_limits<uint32_t>::max()) {
>       |                                                 ^
> external/com_google_absl/absl/synchronization/internal/graphcycles.cc:451:52: error: '::max' has not been declared; did you mean 'std::max'?
>   451 |   if (x->version == std::numeric_limits<uint32_t>::max()) {
>       |                                                    ^~~
>       |                                                    std::max
> In file included from /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/algorithm:62,
>                  from external/com_google_absl/absl/synchronization/internal/graphcycles.cc:38:
> /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/stl_algo.h:3467:5: note: 'std::max' declared here
>  3467 |     max(initializer_list<_Tp> __l, _Compare __comp)
>       |     ^~~
> Target //src:bazel_nojdk failed to build
> INFO: Elapsed time: 538.048s, Critical Path: 73.30s
> INFO: 540 processes: 15 internal, 525 local.
> FAILED: Build did NOT complete successfully
>

**Comment by makingglitches** [comment]
From the look of things this source doesn't persist after the build is terminated.  So this is a little more annoying at first glance than a simple find and replace and some code reorganization.

**Comment by aiuto** [comment]
Are you trying to build bazel from source or trying to build something else with a pre-built bazel?

**Comment by makingglitches** [comment]
I was trying to build bazel from source to build something else because Bazel is not in my distros package repo

Amusingly the generic x86-64 package actually runs on my comp giving me a working install. I didn't expect this to be the case

Now tensorflow build is suffering from some of the same issues and more 

In that case I recognized the sources dragged down into an externals subdirectory are being cached in home/user/.cache/bazel under some convoluted directory structure so that may be the case here and moving forward with the build requires me to make the above mentioned edits and restart the bazel run to fix the issues

My question is why should std c be interpreted so differently ?

Because obviously that is the case between gcc and whatever compiler this build is downloading to be built and whatever is included in the source tree

**Comment by makingglitches** [comment]
Of interest one of the problems in tensorflow is being caused by the same source file related to the com_Google package

**Comment by aiuto** [comment]
I don't know where to help here. This most likely an interaction with whatever compilers that your Fedora distro is selecting vs. what Bazel's and/or Tensor Flow expects.
- You might try on the TensorFlow lists.
- you might try a different compiler (clang instead of gcc)
- you might try a slightly older/newer version of gcc

**Comment by makingglitches** [comment]
There is a solution that could be added to the code itself 
If the compiler is defined as a symbol these files could have only a slight edit made adding a preprocesser statement 
#ifdef enclosing the #include statements that are causing the problem

Why it's defining use prior to definition and on what compiler this would work I have no idea but that's what it's doing

**Comment by aiuto** [comment]
But what code?  This seems to be in multiple upstream and downstream projects. That leads me to think the root case is somewhere between the default ways bazel builds c++ toolchains, the compilers available for fedora 34, and the choices of which of the possible available versions are actually installed.

**Comment by neilg** [comment]
This is related to changes between gcc 10 and gcc 11. If I understand correctly gcc now requires an explicit include of `<limits>` which is where `numeric_limits` is defined.

My system gcc is version 11 and I was getting the same error when running `bazel build //scripts:fish_completion`. Setting `CC=/usr/bin/gcc-10` fixed it.

**Comment by makingglitches** [comment]
@neilg would there be any unforeseen consequences related to this ?

**Comment by neilg** [comment]
@makingglitches I'm sorry, I don't know.

**Comment by makingglitches** [comment]
> @makingglitches I'm sorry, I don't know.

Is the path you provided in all distros that have gcc 11 for backwards compat ?

**Comment by neilg** [comment]
> Is the path you provided in all distros that have gcc 11 for backwards compat ?

I'm using Arch, I think Ubuntu puts the binary in the same place. Off the top of my head I'm not sure about other distros.

**Comment by makingglitches** [comment]
I'll check

**Comment by makingglitches** [comment]
@neilg no dice fedora only has 11. Sigh so code and configure script updates make more sense

**Comment by makingglitches** [comment]
@aiuto if this is not a bot see new issue.

**Comment by makingglitches** [comment]
@janakdr 
at your convenience could you have a look at this repeat bug and consider issue 
https://github.com/bazelbuild/bazel/issues/13635
as the course of action to be taken or link to whoever it is who would make this determination ?

**Comment by janakdr** [comment]
Sorry, I have no knowledge of this area. @philwo

**Comment by philwo** [comment]
We will add support and testing for distributions using gcc 11.x soon. It's currently not yet supported to build Bazel, sorry. 

(Bazel should work fine with gcc 11.x to build other projects - you just can't build Bazel itself yet, but as you found out, our binary releases should work fine.)

**Comment by makingglitches** [comment]
> We will add support and testing for distributions using gcc 11.x soon. It's currently not yet supported to build Bazel, sorry. 
> 
> 
> 
> (Bazel should work fine with gcc 11.x to build other projects - you just can't build Bazel itself yet, but as you found out, our binary releases should work fine.)

There seem to be shared externals with tensorflow which uses bazel to build from source 

This same issue is affecting its build plus a few others

**Comment by makingglitches** [comment]
@philwo that makes me want to ask
Are there really that many differences ? Because this is the only problem I've encountered thus far 

Seems like a high severity low work requirement fix for someone who actually knows the code base

**Comment by davido** [comment]
Last time I checked, there were two problems: one missing include and bump of transitive dependencies.

I fixed both issues with these PRs: https://github.com/bazelbuild/bazel/pull/13537 and https://github.com/bazelbuild/bazel/pull/13536.

The last one is still pending and blocked on: https://github.com/bazelbuild/bazel/issues/13561.

**Comment by makingglitches** [comment]
Yeah I'll pull down dev and check

**Comment by makingglitches** [comment]
@davido 

The precise error that downloading the dist package fixed.
Whats the quick way to get the dev source ready to build ? And OOC why is this necessary ?

>  env EXTRA_BAZEL_ARGS="--host_javabase=@local_jdk//:jdk" bash ./compile.sh
> 
> 🍃  Building Bazel from scratchfind: ‘/usr/share/java/netty-buffer.jar’: No such file or directory
> find: ‘/usr/share/java/netty-codec-dns.jar’: No such file or directory
> find: ‘/usr/share/java/netty-codec-haproxy.jar’: No such file or directory
> find: ‘/usr/share/java/netty-codec-http2.jar’: No such file or directory
> find: ‘/usr/share/java/netty-codec-http.jar’: No such file or directory
> find: ‘/usr/share/java/netty-codec.jar’: No such file or directory
> find: ‘/usr/share/java/netty-codec-memcache.jar’: No such file or directory
> find: ‘/usr/share/java/netty-codec-mqtt.jar’: No such file or directory
> find: ‘/usr/share/java/netty-codec-redis.jar’: No such file or directory
> find: ‘/usr/share/java/netty-codec-smtp.jar’: No such file or directory
> find: ‘/usr/share/java/netty-codec-socks.jar’: No such file or directory
> find: ‘/usr/share/java/netty-codec-stomp.jar’: No such file or directory
> find: ‘/usr/share/java/netty-common.jar’: No such file or directory
> find: ‘/usr/share/java/netty-handler.jar’: No such file or directory
> find: ‘/usr/share/java/netty-handler-proxy.jar’: No such file or directory
> find: ‘/usr/share/java/netty-resolver-dns.jar’: No such file or directory
> find: ‘/usr/share/java/netty-resolver-dns-native-macos.jar’: No such file or directory
> find: ‘/usr/share/java/netty-resolver.jar’: No such file or directory
> find: ‘/usr/share/java/netty-transport.jar’: No such file or directory
> find: ‘/usr/share/java/netty-transport-native-epoll.jar’: No such file or directory
> find: ‘/usr/share/java/netty-transport-native-kqueue.jar’: No such file or directory
> find: ‘/usr/share/java/netty-transport-native-unix-common.jar’: No such file or directory
> find: ‘/usr/share/java/netty-transport-sctp.jar’: No such file or directory
> find: ‘/usr/share/java/gson.jar’: No such file or directory
> find: ‘/usr/share/java/tomcat9-annotations-api.jar’: No such file or directory
> find: ‘/usr/share/java/tomcat9-annotations-api.jar’: No such file or directory
> find: ‘/usr/share/java/java-allocation-instrumenter.jar’: No such file or directory
> find: ‘/usr/share/java/protobuf.jar’: No such file or directory
> find: ‘/usr/share/java/protobuf-util.jar’: No such file or directory
> find: ‘/usr/share/java/protobuf.jar’: No such file or directory
> find: ‘/usr/share/java/protobuf-util.jar’: No such file or directory
> find: ‘/usr/share/java/commons-collections3.jar’: No such file or directory
> find: ‘/usr/share/java/commons-lang.jar’: No such file or directory
> find: ‘/usr/share/java/commons-compress.jar’: No such file or directory
> find: ‘/usr/share/java/commons-logging.jar’: No such file or directory
> find: ‘/usr/share/java/commons-pool2.jar’: No such file or directory
> find: ‘/usr/share/java/velocity.jar’: No such file or directory
> find: ‘/usr/share/java/asm.jar’: No such file or directory
> find: ‘/usr/share/java/jackson-core.jar’: No such file or directory
> find: ‘/usr/share/java/jcip-annotations.jar’: No such file or directory
> find: ‘/usr/share/java/jcip-annotations.jar’: No such file or directory
> find: ‘/usr/share/java/jsr305.jar’: No such file or directory
> find: ‘/usr/share/java/jsr305.jar’: No such file or directory
> find: ‘/usr/share/java/netty-tcnative.jar’: No such file or directory
> find: ‘/usr/share/java/javapoet.jar’: No such file or directory
> find: ‘/usr/share/java/jaxb-api.jar’: No such file or directory
> find: ‘/usr/share/java/xz.jar’: No such file or directory
> find: ‘/usr/share/java/geronimo-annotation-1.3-spec.jar’: No such file or directory
> find: ‘/usr/share/java/com.android.tools.common.jar’: No such file or directory
> find: ‘/usr/share/java/com.android.tools.repository.jar’: No such file or directory
> find: ‘/usr/share/java/com.android.tools.layoutlib.layoutlib-api.jar’: No such file or directory
> find: ‘/usr/share/java/guava.jar’: No such file or directory
> find: ‘/usr/share/java/guava.jar’: No such file or directory
> find: ‘/usr/share/java/org.jacoco.agent.jar’: No such file or directory
> find: ‘/usr/share/java/org.jacoco.agent.jar’: No such file or directory
> find: ‘/usr/share/java/org.jacoco.core.jar’: No such file or directory
> find: ‘/usr/share/java/org.jacoco.core.jar’: No such file or directory
> find: ‘/usr/share/java/org.jacoco.core.jar’: No such file or directory
> find: ‘/usr/share/java/org.jacoco.core.jar’: No such file or directory
> find: ‘/usr/share/java/org.jacoco.report.jar’: No such file or directory
> find: ‘/usr/share/java/org.jacoco.report.jar’: No such file or directory
> find: ‘/usr/share/java/org.jacoco.agent.jar’: No such file or directory
> find: ‘/usr/share/java/org.jacoco.agent.jar’: No such file or directory
> find: ‘/usr/share/java/auto-common.jar’: No such file or directory
> find: ‘/usr/share/java/auto-service.jar’: No such file or directory
> find: ‘/usr/share/java/auto-service-annotations.jar’: No such file or directory
> find: ‘/usr/share/java/escapevelocity.jar’: No such file or directory
> find: ‘/usr/share/java/auto-value.jar’: No such file or directory
> find: ‘/usr/share/java/auto-value-annotations.jar’: No such file or directory
> find: ‘/usr/share/java/auto-common.jar’: No such file or directory
> find: ‘/usr/share/java/auto-value.jar’: No such file or directory
> find: ‘/usr/share/java/auto-value-annotations.jar’: No such file or directory
> find: ‘/usr/share/java/escapevelocity.jar’: No such file or directory
> find: ‘/usr/share/java/javapoet.jar’: No such file or directory
> find: ‘/usr/share/java/google-api-client.jar’: No such file or directory
> find: ‘/usr/share/java/google-api-client-jackson2.jar’: No such file or directory
> find: ‘/usr/share/java/google-http-client.jar’: No such file or directory
> find: ‘/usr/share/java/google-http-client-jackson2.jar’: No such file or directory
> find: ‘/usr/share/java/google-auth-library-oauth2-http.jar’: No such file or directory
> find: ‘/usr/share/java/google-auth-library-credentials.jar’: No such file or directory
> find: ‘/usr/share/java/error-prone-annotations.jar’: No such file or directory
> find: ‘/usr/share/java/error-prone-type-annotations.jar’: No such file or directory
> find: ‘/usr/share/java/error-prone-annotations.jar’: No such file or directory
> find: ‘/usr/share/java/java-diff-utils.jar’: No such file or directory
> find: ‘/usr/share/java/opencensus-api.jar’: No such file or directory
> find: ‘/usr/share/java/opencensus-contrib-grpc-metrics.jar’: No such file or directory
> find: ‘/usr/share/java/perfmark-api.jar’: No such file or directory
> find: ‘/usr/share/java/flogger.jar’: No such file or directory
> find: ‘/usr/share/java/flogger-system-backend.jar’: No such file or directory
> find: ‘/usr/share/java/google-extensions.jar’: No such file or directory
> find: ‘/usr/share/java/flogger.jar’: No such file or directory
> find: ‘/usr/share/java/flogger-system-backend.jar’: No such file or directory
> find: ‘/usr/share/java/google-extensions.jar’: No such file or directory
> *find: ‘/usr/share/java/checker-qual.jar’: No such file or directory
> find: ‘/usr/share/java/grpc-api.jar’: No such file or directory
> find: ‘/usr/share/java/grpc-auth.jar’: No such file or directory
> find: ‘/usr/share/java/grpc-context.jar’: No such file or directory
> find: ‘/usr/share/java/grpc-core.jar’: No such file or directory
> find: ‘/usr/share/java/grpc-netty.jar’: No such file or directory
> find: ‘/usr/share/java/grpc-protobuf.jar’: No such file or directory
> find: ‘/usr/share/java/grpc-protobuf-lite.jar’: No such file or directory
> find: ‘/usr/share/java/grpc-stub.jar’: No such file or directory
> find: ‘/usr/share/java/hamcrest-core.jar’: No such file or directory
> find: ‘/usr/share/java/junit4.jar’: No such file or directory
> 
> ERROR: Must specify PROTOC if not bootstrapping from the distribution artifact
> 
> --------------------------------------------------------------------------------
> NOTE: This failure is likely occuring if you are trying to bootstrap bazel from
> a developer checkout. Those checkouts do not include the generated output of
> the protoc compiler (as we prefer not to version generated files).
> 
> * To build a developer version of bazel, do
> 
>     bazel build //src:bazel
> 
> * To bootstrap your first bazel binary, please download a dist archive from our
>   release page at https://github.com/bazelbuild/bazel/releases and run
>   compile.sh on the unpacked archive.
> 
> The full install instructions to install a release version of bazel can be found
> at https://docs.bazel.build/install-compile-source.html
> For a rationale, why the bootstrap process is organized in this way, see
> https://bazel.build/designs/2016/10/11/distribution-artifact.html
> --------------------------------------------------------------------------------
>

**Comment by makingglitches** [comment]
Is it saying I HAVE to do a build using bazel ? Because while I have a binary dist installed for tensorflow,  wasn't the steps I followed to generate the errors this post is about.

**Comment by makingglitches** [comment]
@davido  nogo

> 
> [john@localhost bazel-master]$  bazel build //src:bazel
> Starting local Bazel server and connecting to it...
> DEBUG: /home/john/.cache/bazel/_bazel_john/9ade6b4a30d3c956d633ebb726c4c137/external/bazel_toolchains/rules/rbe_repo/version_check.bzl:68:14: 
> Current running Bazel is ahead of bazel-toolchains repo. Please update your pin to bazel-toolchains repo in your WORKSPACE file.
> DEBUG: /home/john/.cache/bazel/_bazel_john/9ade6b4a30d3c956d633ebb726c4c137/external/bazel_toolchains/rules/rbe_repo/checked_in.bzl:103:14: rbe_ubuntu1804_java11 not using checked in configs as detect_java_home was set to True 
> DEBUG: /home/john/.cache/bazel/_bazel_john/9ade6b4a30d3c956d633ebb726c4c137/external/bazel_toolchains/rules/rbe_repo/version_check.bzl:68:14: 
> Current running Bazel is ahead of bazel-toolchains repo. Please update your pin to bazel-toolchains repo in your WORKSPACE file.
> DEBUG: /home/john/.cache/bazel/_bazel_john/9ade6b4a30d3c956d633ebb726c4c137/external/bazel_toolchains/rules/rbe_repo/checked_in.bzl:103:14: rbe_ubuntu1604_java8 not using checked in configs as detect_java_home was set to True 
> INFO: Analyzed target //src:bazel (334 packages loaded, 9039 targets configured).
> INFO: Found 1 target...
> INFO: From Compiling src/google/protobuf/message_lite.cc [for host]:
> In file included from /usr/include/string.h:519,
>                  from external/com_google_protobuf/src/google/protobuf/stubs/port.h:39,
>                  from external/com_google_protobuf/src/google/protobuf/stubs/macros.h:34,
>                  from external/com_google_protobuf/src/google/protobuf/stubs/common.h:46,
>                  from external/com_google_protobuf/src/google/protobuf/message_lite.h:45,
>                  from external/com_google_protobuf/src/google/protobuf/message_lite.cc:36:
> In function 'void* memcpy(void*, const void*, size_t)',
>     inlined from 'google::protobuf::uint8* google::protobuf::io::EpsCopyOutputStream::WriteRaw(const void*, int, google::protobuf::uint8*)' at external/com_google_protobuf/src/google/protobuf/io/coded_stream.h:699:16,
>     inlined from 'virtual google::protobuf::uint8* google::protobuf::internal::ImplicitWeakMessage::_InternalSerialize(google::protobuf::uint8*, google::protobuf::io::EpsCopyOutputStream*) const' at external/com_google_protobuf/src/google/protobuf/implicit_weak_message.h:85:28,
>     inlined from 'bool google::protobuf::MessageLite::SerializePartialToZeroCopyStream(google::protobuf::io::ZeroCopyOutputStream*) const' at external/com_google_protobuf/src/google/protobuf/message_lite.cc:419:30:
> /usr/include/bits/string_fortified.h:29:33: warning: 'void* __builtin___memcpy_chk(void*, const void*, long unsigned int, long unsigned int)' specified size between 18446744071562067968 and 18446744073709551615 exceeds maximum object size 9223372036854775807 [-Wstringop-overflow=]
>    29 |   return __builtin___memcpy_chk (__dest, __src, __len,
>       |          ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
>    30 |                                  __glibc_objsize0 (__dest));
>       |                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~
> INFO: From Compiling src/google/protobuf/message_lite.cc:
> In file included from /usr/include/string.h:519,
>                  from external/com_google_protobuf/src/google/protobuf/stubs/port.h:39,
>                  from external/com_google_protobuf/src/google/protobuf/stubs/macros.h:34,
>                  from external/com_google_protobuf/src/google/protobuf/stubs/common.h:46,
>                  from external/com_google_protobuf/src/google/protobuf/message_lite.h:45,
>                  from external/com_google_protobuf/src/google/protobuf/message_lite.cc:36:
> In function 'void* memcpy(void*, const void*, size_t)',
>     inlined from 'google::protobuf::uint8* google::protobuf::io::EpsCopyOutputStream::WriteRaw(const void*, int, google::protobuf::uint8*)' at external/com_google_protobuf/src/google/protobuf/io/coded_stream.h:699:16,
>     inlined from 'virtual google::protobuf::uint8* google::protobuf::internal::ImplicitWeakMessage::_InternalSerialize(google::protobuf::uint8*, google::protobuf::io::EpsCopyOutputStream*) const' at external/com_google_protobuf/src/google/protobuf/implicit_weak_message.h:85:28,
>     inlined from 'bool google::protobuf::MessageLite::SerializePartialToZeroCopyStream(google::protobuf::io::ZeroCopyOutputStream*) const' at external/com_google_protobuf/src/google/protobuf/message_lite.cc:419:30:
> /usr/include/bits/string_fortified.h:29:33: warning: 'void* __builtin___memcpy_chk(void*, const void*, long unsigned int, long unsigned int)' specified size between 18446744071562067968 and 18446744073709551615 exceeds maximum object size 9223372036854775807 [-Wstringop-overflow=]
>    29 |   return __builtin___memcpy_chk (__dest, __src, __len,
>       |          ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
>    30 |                                  __glibc_objsize0 (__dest));
>       |                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~
> ERROR: /home/john/.cache/bazel/_bazel_john/9ade6b4a30d3c956d633ebb726c4c137/external/com_google_absl/absl/synchronization/BUILD.bazel:30:11: Compiling absl/synchronization/internal/graphcycles.cc failed: (Exit 1): gcc failed: error executing command /usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer '-std=c++0x' ... (remaining 34 argument(s) skipped)
> 
> Use --sandbox_debug to see verbose messages from the sandbox
> external/com_google_absl/absl/synchronization/internal/graphcycles.cc: In member function 'void absl::lts_2020_02_25::synchronization_internal::GraphCycles::RemoveNode(void*)':
> external/com_google_absl/absl/synchronization/internal/graphcycles.cc:451:26: error: 'numeric_limits' is not a member of 'std'
>   451 |   if (x->version == std::numeric_limits<uint32_t>::max()) {
>       |                          ^~~~~~~~~~~~~~
> external/com_google_absl/absl/synchronization/internal/graphcycles.cc:451:49: error: expected primary-expression before '>' token
>   451 |   if (x->version == std::numeric_limits<uint32_t>::max()) {
>       |                                                 ^
> external/com_google_absl/absl/synchronization/internal/graphcycles.cc:451:52: error: '::max' has not been declared; did you mean 'std::max'?
>   451 |   if (x->version == std::numeric_limits<uint32_t>::max()) {
>       |                                                    ^~~
>       |                                                    std::max
> In file included from /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/algorithm:62,
>                  from external/com_google_absl/absl/synchronization/internal/graphcycles.cc:38:
> /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/stl_algo.h:3467:5: note: 'std::max' declared here
>  3467 |     max(initializer_list<_Tp> __l, _Compare __comp)
>       |     ^~~
> Target //src:bazel failed to build
> Use --verbose_failures to see the command lines of failed build steps.
> INFO: Elapsed time: 1028.749s, Critical Path: 62.24s
> INFO: 544 processes: 17 internal, 527 linux-sandbox.
> FAILED: Build did NOT complete successfully
>

**Comment by philwo** [comment]
I will look into fixing this, but I'll be busy with other stuff for the next 2 weeks.

**Comment by sventiffe** [comment]
This issue did not receive an update for a while. @philwo do we still consider this a P1 and if so please update

**Comment by philwo** [comment]
Yes, this is a P1. There is no update yet.

**Comment by philwo** [comment]
@meteorcloudy Could you help me look into this? Currently Bazel cannot be compiled on newer Linux distributions due to this issue. I'll prepare a new Ubuntu container that we can use to test this with, then we could add it as "soft_fail" to our CI.

**Comment by meteorcloudy** [comment]
No problem, do you have a way to reproduce now?

**Comment by philwo** [comment]
Yes, if you try to build Bazel on a recent Linux distribution that uses gcc 11.x, this will fail to build. Ubuntu 21.04 uses gcc-10 as default, but has gcc-11 optionally. Arch Linux and Fedora 34 already use gcc 11.x.

IIRC, the latest version of Clang is affected too, when I tried to built Bazel using it on my Arch Linux machine.

**Comment by makingglitches** [comment]
@philwo   gcc 10 is then apparently displaying non standard behavior as it is expecting header files to be included when not specifically included.  thats the isssue that is occurring.

there is a chain of inclusion.

if i include from an h file that gets included, then all of that becomes available to the source file.
if i include only in a source .c or .cpp file it is to be local to that source file.
that is what is causing this issue.
